### PR TITLE
feature/PT-6/order-history

### PR DESCRIPTION
**Pull request by dungeon-slave**
add order history feature
fixes

https://github.com/dungeon-slave/food_app/assets/70040363/f15bce32-d5fd-4e13-a23a-d5a6571a5ca5

### CODE REVIEW

File: data/lib/repositories/authentication_repository_impl.dart [hunk 1]
```diff
@@ -36,5 +36,5 @@ class AuthenticationRepositoryImpl implements AuthenticationRepository {
       await _hiveProvider.saveUserId(userId);
 
   @override
-  bool checkUser() => _hiveProvider.fetchUserId() != null ? true : false;
+  bool checkUser() => _hiveProvider.fetchUserId() != '' ? true : false;
```
Comment (ibondarovich) [inline data/lib/repositories/authentication_repository_impl.dart]: naming

File: features/dishes_menu/lib/src/ui/dish_item.dart [hunk 2]
```diff
@@ -8,10 +8,10 @@ import 'package:core_ui/core_ui.dart';
 class DishItem extends StatefulWidget {
   final DishModel _model;
 
-  void Function() _addToCartHandler(BuildContext context) {
-    return () =>
-        BlocProvider.of<MenuBloc>(context).add(AddDishEvent(model: _model));
-  }
+  void Function() _addToCartHandler(BuildContext context) =>
+      () => BlocProvider.of<MenuBloc>(context).add(
+            AddDishEvent(model: _model)
```
Comment (ibondarovich) [inline features/dishes_menu/lib/src/ui/dish_item.dart]: ,

File: features/home_screen/lib/src/ui/home_screen.dart [hunk 3]
```diff
@@ -60,37 +60,39 @@ class HomeScreen extends StatelessWidget {
...
                       label: AppConstants.orderHistoryTitle,
                     ),
                     BottomNavigationBarItem(
                       activeIcon: BlocBuilder<HomeScreenBloc, HomeScreenState>(
                         builder: (context, state) {
                           return CartCountBadge(
                             count: state.count,
-                            themeIcon: AppIcons.selectedShoppingCart,
+                            themeIcon:
+                                const AppIcon(AppIconsData.selectedShoppingCart),
```
Comment (ibondarovich) [inline features/home_screen/lib/src/ui/home_screen.dart]: ,

File: features/home_screen/lib/src/ui/home_screen.dart [hunk 4]
```diff
@@ -60,37 +60,39 @@ class HomeScreen extends StatelessWidget {
...
                           );
                         },
                       ),
                       icon: BlocBuilder<HomeScreenBloc, HomeScreenState>(
                         builder: (context, state) {
                           return CartCountBadge(
                             count: state.count,
-                            themeIcon: AppIcons.unselectedShoppingCart,
+                            themeIcon:
+                                const AppIcon(AppIconsData.unselectedShoppingCart),
```
Comment (ibondarovich) [inline features/home_screen/lib/src/ui/home_screen.dart]: ,

File: features/order_history/lib/src/ui/components/main_info.dart [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            thickness: AppConstants.orderDividerThickness,
+            endIndent: AppDimens.padding15,
+            color: Theme.of(context).indicatorColor,
+          ),
+        ),
+        Container(
+          padding: const EdgeInsets.only(bottom: AppDimens.padding10),
+          alignment: Alignment.centerLeft,
+          child: Text(
+            '${AppConstants.date}: ${_model.dateTime.day}.${_model.dateTime.month}.${_model.dateTime.year}',
```
Comment (ibondarovich) [inline features/order_history/lib/src/ui/components/main_info.dart:39]: formatting?

File: features/order_history/lib/src/ui/components/main_info.dart [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            style: AppFonts.normal22,
+          ),
+        ),
+        Container(
+          padding: const EdgeInsets.only(bottom: AppDimens.padding10),
+          alignment: Alignment.centerLeft,
+          child: Text(
+            '${AppConstants.time}: ${_model.dateTime.hour}:${_model.dateTime.minute}',
```
Comment (ibondarovich) [inline features/order_history/lib/src/ui/components/main_info.dart:47]: same

File: features/order_history/lib/src/ui/components/main_info.dart [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            style: AppFonts.normal22,
+          ),
+        ),
+        Container(
+          padding: const EdgeInsets.only(bottom: AppDimens.margin5),
+          alignment: Alignment.centerLeft,
+          child: Text(
+            '${AppConstants.total}: ${_model.price.toStringAsFixed(AppConstants.priceSize)}\$',
```
Comment (ibondarovich) [inline features/order_history/lib/src/ui/components/main_info.dart:55]: same

File: features/order_history/lib/src/ui/components/order_history_item.dart [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                          itemBuilder: (context, index) {
+                            return Container(
+                              margin: const EdgeInsets.only(
+                                top: AppDimens.margin5,
+                                bottom: AppDimens.margin5,
+                                left: AppDimens.margin5,
+                              ),
+                              alignment: Alignment.centerLeft,
+                              child: Text(
+                                '${index + 1}. ${widget._model.products[index].dishModel.name} x ${widget._model.products[index].count}',
```
Comment (ibondarovich) [inline features/order_history/lib/src/ui/components/order_history_item.dart:81]: same

File: features/shopping_cart/lib/src/ui/shopping_cart.dart [hunk 9]
```diff
@@ -104,15 +109,29 @@ class ShoppingCartScreen extends StatelessWidget {
                   ],
                 ),
                 SizedBox(
-                  width: AppDimens.size400, //FIXME make adaptive size
+                  width: kTextTabBarHeight * 8,
                   child: BlocBuilder<ShoppingCartBloc, ShoppingCartState>(
                     builder: (context, state) {
```
Comment (ibondarovich) [inline features/shopping_cart/lib/src/ui/shopping_cart.dart]: type

File: features/shopping_cart/lib/src/ui/shopping_cart.dart [hunk 10]
```diff
@@ -30,7 +32,7 @@ class ShoppingCartScreen extends StatelessWidget {
                   builder: (context, state) {
```
Comment (ibondarovich) [inline features/shopping_cart/lib/src/ui/shopping_cart.dart]: type

File: features/sign_in_screen/lib/src/ui/sign_in_screen.dart [hunk 11]
```diff
@@ -28,45 +40,53 @@ class SignInScreen extends StatelessWidget {
...
-                AppConstants.signInPageTitle,
-                style: AppFonts.normal30.copyWith(
-                  color: Theme.of(context).primaryColor,
+        body: BlocConsumer<SignInBloc, SignInState>(
+          listener: (context, state) {
+            if (state.errorMessage.isNotEmpty) {
+              showAppSnackBar(context: context, title: state.errorMessage);
+            }
+          },
+          builder: (context, state) {
```
Comment (ibondarovich) [inline features/sign_in_screen/lib/src/ui/sign_in_screen.dart]: type

File: features/sign_up_screen/lib/src/ui/sign_up_screen.dart [hunk 12]
```diff
@@ -8,13 +8,23 @@ import 'package:navigation/navigation.dart';
...
+class SignUpScreen extends StatefulWidget {
+  const SignUpScreen({super.key});
+
+  @override
+  State<SignUpScreen> createState() => _SignUpState();
+}
+
+class _SignUpState extends State<SignUpScreen> {
   final _emailController = TextEditingController();
   final _passwordController = TextEditingController();
```
Comment (ibondarovich) [inline features/sign_up_screen/lib/src/ui/sign_up_screen.dart]: type

File: features/sign_up_screen/lib/src/ui/sign_up_screen.dart [hunk 13]
```diff
@@ -27,45 +37,53 @@ class SignUpScreen extends StatelessWidget {
...
-                AppConstants.signUpPageTitle,
-                style: AppFonts.normal30.copyWith(
-                  color: Theme.of(context).primaryColor,
+        body: BlocConsumer<SignUpBloc, SignUpState>(
+          listener: (context, state) {
+            if (state.errorMessage.isNotEmpty) {
+              showAppSnackBar(context: context, title: state.errorMessage);
+            }
+          },
+          builder: (context, state) {
```
Comment (ibondarovich) [inline features/sign_up_screen/lib/src/ui/sign_up_screen.dart]: type

**Review by ibondarovich — CHANGES_REQUESTED** [review]

File: core_ui/lib/src/theme/app_constants.dart [hunk 14]
```diff
@@ -1,15 +1,28 @@
...
   static const String menuCollection = 'dishes_menu';
+  static const String usersCollection = 'users';
+  static const String userOrdersCollection = 'orders';
+  static const String userCartCollection = 'cart';
 
   static const String contactUs = 'Contact us';
+  static const String showDetails = 'Show details';
+  static const String hideDetails = 'Hide details';
+
+  static const emptyList = 'It\'s empty here...';
```
Comment (Alexey-Lebedzeu) [inline core_ui/lib/src/theme/app_constants.dart]: type

File: core_ui/lib/src/theme/app_constants.dart [hunk 15]
```diff
@@ -1,15 +1,28 @@
...
 
-  static const String telegramLink = 'https://pub.dev/packages/url_launcher/example';
-  static const String instagramLink = 'https://api.flutter.dev/flutter/material/Icons-class.html';
+  static const String telegramLink =
+      'https://pub.dev/packages/url_launcher/example';
+  static const String instagramLink =
+      'https://api.flutter.dev/flutter/material/Icons-class.html';
 
   static const int priceSize = 2;
+  static const int cartSnackBarDuration = 2;
```
Comment (Alexey-Lebedzeu) [inline core_ui/lib/src/theme/app_constants.dart]: separate

File: data/lib/providers/remote/firebase_provider.dart [hunk 16]
```diff
@@ -1,31 +1,38 @@
 import 'package:core/core.dart' show FirebaseFirestore, QueryDocumentSnapshot;
 import 'package:core_ui/core_ui.dart' show AppConstants;
-import 'package:data/entities/dish_entity/dish_entity.dart';
 import 'package:data/entities/dish_type_enity/dish_type_entity.dart';
+import 'package:data/entities/order_history_entity/order_entity.dart';
 
 class FirebaseProvider {
-  Future<List<DishTypeEntity>> fetchMenu() async {
-    List<DishTypeEntity> dishesTypes = <DishTypeEntity>[];
+  Future<List<DishTypeEntity>> fetchMenu() async =>
```
Comment (Alexey-Lebedzeu) [inline data/lib/providers/remote/firebase_provider.dart]: refactor this

File: data/lib/repositories/cart_repository_impl.dart [hunk 17]
```diff
@@ -34,7 +45,15 @@ class CartRepositoryImpl implements CartRepository {
...
   @override
   Future<void> changeItemCount(CartItemModel item) =>
-      _hiveProvider.changeItemCount(
-        CartItemMapper.toEntity(item),
-      );
+      _hiveProvider.changeItemCount(CartItemMapper.toEntity(item));
+
+  @override
+  Future<void> sendOrder(OrderModel item) async {
+    await _networkService.checkConnection()
```
Comment (Alexey-Lebedzeu) [inline data/lib/repositories/cart_repository_impl.dart]: refactor this to safe request

File: domain/lib/repositories/authentication_repository.dart [hunk 18]
```diff
@@ -2,14 +2,10 @@ import 'package:domain/models/authentication/email_sign_in_model.dart';
...
-    required EmailSignUpModel data,
-  });
-  Future<String> emailSignIn({
-    required EmailSignInModel data,
-  });
+  Future<String> emailSignUp(EmailSignUpModel data);
+  Future<String> emailSignIn(EmailSignInModel data);
   Future<String> googleSignIn();
   Future<void> signOut();
   Future<void> saveUser(String userId);
```
Comment (Alexey-Lebedzeu) [inline domain/lib/repositories/authentication_repository.dart]: replace

File: domain/lib/usecase/shopping_cart/send_order_usecase.dart [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+
+class SendOrderUseCase implements AsyncUseCase<OrderModel, void> {
+  final CartRepository _cartRepository;
+
+  const SendOrderUseCase({
+    required CartRepository cartRepository,
+  }) : _cartRepository = cartRepository;
+
+  @override
+  Future<void> execute(OrderModel input) async =>
```
Comment (Alexey-Lebedzeu) [inline domain/lib/usecase/shopping_cart/send_order_usecase.dart]: remove async

File: features/detailed_page/lib/src/ui/detailed_page_screen.dart [hunk 20]
```diff
@@ -30,8 +30,11 @@ class DetailedPageScreen extends StatelessWidget {
...
               child: IconButton(
                 iconSize: AppConstants.deleteIconSize,
-                icon: AppIcons.closeDetailedPage,
-            color: Theme.of(context).primaryColor,
+                icon: const AppIcon(
+                  AppIconsData.closeDetailedPage,
+                  size: AppDimens.size40,
+                ),
+                color: Theme.of(context).primaryColor,
                 onPressed: () => context.router.pop(),
```
Comment (Alexey-Lebedzeu) [inline features/detailed_page/lib/src/ui/detailed_page_screen.dart]: to bloc

File: features/order_history/lib/src/ui/components/order_history_item.dart [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+                  AppTextButton(
+                    text: _isVisible
+                        ? AppConstants.hideDetails
+                        : AppConstants.showDetails,
+                    handler: _changeVisibility,
+                    textStyle: AppFonts.normal24.copyWith(
+                      color: Theme.of(context).indicatorColor,
+                    ),
+                  ),
+                  _isVisible
```
Comment (Alexey-Lebedzeu) [inline features/order_history/lib/src/ui/components/order_history_item.dart]: if

File: features/order_history/lib/src/ui/order_history.dart [hunk 22]
```diff
@@ -1,20 +1,61 @@
...
+            backgroundColor: Theme.of(context).scaffoldBackgroundColor,
+            title: Text(
+              AppConstants.orderHistoryTitle,
+              style: AppFonts.normal30.copyWith(
+                color: Theme.of(context).primaryColor,
+              ),
+            ),
+          ),
+          body: BlocBuilder<OrderHistoryBloc, OrderHistoryState>(
+            builder: (context, state) {
```
Comment (Alexey-Lebedzeu) [inline features/order_history/lib/src/ui/order_history.dart]: type

File: features/order_history/lib/src/ui/order_history.dart [hunk 23]
```diff
@@ -1,20 +1,61 @@
...
   Widget build(BuildContext context) {
-    return Container(
-      color: Theme.of(context).scaffoldBackgroundColor,
-      child: Center(
-        child: Text(
-          AppConstants.orderHistoryTitle,
-          style: AppFonts.normal25.copyWith(
-            color: Theme.of(context).indicatorColor,
+    return BlocProvider(
+      create: (context) => OrderHistoryBloc(
```
Comment (Alexey-Lebedzeu) [inline features/order_history/lib/src/ui/order_history.dart]: type

File: features/shopping_cart/lib/src/bloc/shopping_cart/shopping_cart_bloc.dart [hunk 24]
```diff
@@ -65,34 +70,70 @@ class ShoppingCartBloc extends Bloc<ShoppingCartEvent, ShoppingCartState> {
...
+      ),
+    );
   }
 
   Future<void> _incrementItem(
-      IncrementEvent event, Emitter<ShoppingCartState> emit) async {
+    IncrementEvent event,
+    Emitter<ShoppingCartState> emit,
+  ) async {
     int index = state.items.indexWhere((CartItemModel element) =>
```
Comment (Alexey-Lebedzeu) [inline features/shopping_cart/lib/src/bloc/shopping_cart/shopping_cart_bloc.dart]: remove it and impl increment and decrement in model

File: features/shopping_cart/lib/src/bloc/shopping_cart/shopping_cart_bloc.dart [hunk 25]
```diff
@@ -65,34 +70,70 @@ class ShoppingCartBloc extends Bloc<ShoppingCartEvent, ShoppingCartState> {
...
-  void _checkout(CheckoutEvent event, Emitter<ShoppingCartState> emit) {
-    //TODO implement checkout
+  Future<void> _checkout(
+    CheckoutEvent event,
+    Emitter<ShoppingCartState> emit,
+  ) async {
+    try {
+      await _sendOrderUseCase.execute(
+        OrderModel(
+          id: DateTime.now().microsecondsSinceEpoch.toString(),
```
Comment (Alexey-Lebedzeu) [inline features/shopping_cart/lib/src/bloc/shopping_cart/shopping_cart_bloc.dart]: UUID

File: features/shopping_cart/lib/src/ui/shopping_cart.dart [hunk 26]
```diff
@@ -104,15 +109,29 @@ class ShoppingCartScreen extends StatelessWidget {
                   ],
                 ),
                 SizedBox(
-                  width: AppDimens.size400, //FIXME make adaptive size
+                  width: kTextTabBarHeight * 8,
```
Comment (Alexey-Lebedzeu) [inline features/shopping_cart/lib/src/ui/shopping_cart.dart]: refactor this

File: features/sign_in_screen/lib/src/bloc/sign_in_bloc.dart [hunk 27]
```diff
@@ -33,47 +33,57 @@ class SignInBloc extends Bloc<LoginScreenEvent, LoginScreenState> {
         _checkUserUseCase = checkUserUseCase,
         _authService = authService,
         _appRouter = appRouter,
-        super(const LoginScreenState(errorMessage: '')) {
+        super(
+          const SignInState(),
```
Comment (Alexey-Lebedzeu) [inline features/sign_in_screen/lib/src/bloc/sign_in_bloc.dart]: ,?

File: features/sign_up_screen/lib/src/bloc/sign_up_bloc.dart [hunk 28]
```diff
@@ -37,10 +37,24 @@ class SignUpBloc extends Bloc<SignUpEvent, SignUpState> {
...
-    _appRouter.replace(const HomeRoute());
-    //TODO implement state change
+    emit(
+      state.copyWith(
+        isLoading: true,
+        errorMessage: '',
+      ),
+    );
+    try {
+      String userId = await _emailSignUpUseCase.execute(event.data);
```
Comment (Alexey-Lebedzeu) [inline features/sign_up_screen/lib/src/bloc/sign_up_bloc.dart]: final

**Review by Alexey-Lebedzeu — CHANGES_REQUESTED** [review]

**Pull request merged**